### ISSUE TITLE
(TypeScript) Generic is not picked up properly

### ISSUE THREAD

**Issue by ackvf** [highlightjs/highlight.js#3603]
**Describe the issue**

**Which language seems to have the issue?**

typescript, no auto-detection

**Are you using `highlight` or `highlightAuto`?**
I don't know what this means.

**Sample Code to Reproduce**

I tried to reproduce the issue using your jsfiddle, but the typescript generic is not even displaying there ¯\\_(ツ)_/¯.  
~~https://jsfiddle.net/x3nzp2rf/~~  
updated: https://jsfiddle.net/sm706vyd/2/

I have this code on StackOverflow.

meta: https://meta.stackoverflow.com/questions/420116/typescript-syntax-highlight-with-generic-not-picked-up-properly

first observed issue here: https://stackoverflow.com/questions/72794594/typescript-interface-merging-to-make-an-external-libs-interface-property-more-s/73527814#73527814

    ```lang-ts
    import { useWeb3React as useWeb3React_ } from '@web3-react/core'
    
    export const useWeb3React: <T = any>(key?: string) => Modify<
      ReturnType<typeof useWeb3React_<T>>,
      { chainId: SupportedChainIds }
    > = useWeb3React_ as any
    
    declare global {
      type SupportedChainIds = 1 | 4
    }
    ```

Which produces incorrect syntax highlight

``` lang-ts
import { useWeb3React as useWeb3React_ } from '@web3-react/core'

export const useWeb3React: <T = any>(key?: string) => Modify<
  ReturnType<typeof useWeb3React_<T>>,
  { chainId: SupportedChainIds }
> = useWeb3React_ as any

declare global {
  type SupportedChainIds = 1 | 4
}
```

After a tweak I achieve correct syntax highlight

``` lang-ts
import { useWeb3React as useWeb3React_ } from '@web3-react/core'

export const useWeb3React: < T = any >(key?: string) => Modify<
  ReturnType<typeof useWeb3React_<T>>,
  { chainId: SupportedChainIds }
> = useWeb3React_ as any

declare global {
  type SupportedChainIds = 1 | 4
}
```

This is the tweak:
```
- export const useWeb3React: <T = any>(key?: string) => Modify<
+ export const useWeb3React: < T = any >(key?: string) => Modify<
```

**Expected behavior**

(without the spaces around `< T = any >` - correct is: `<T = any>`)
![](https://i.stack.imgur.com/LOQbR.png)

**Actual behavior**

![](https://i.stack.imgur.com/JqOmd.png)

**Additional context**

**Comment by joshgoebel** [comment]
> I tried to reproduce the issue using your jsfiddle, but the typescript generic is not even displaying there ¯\(ツ)/¯.

Because you haven't properly escaped `<` and `>` in your HTML, etc.

**Comment by joshgoebel** [comment]
Not sure I follow this `<T = ` syntax, could a simpler example be given?  I believe the problem is this is detected as HTML starting and the end of the HTML can't ever be found, so the highlighting is broken.  JSX and types are very hard to parse without a real parser.

**Comment by joshgoebel** [comment]
#3604 might be a reasonable fix. Thoughts?

**Comment by ackvf** [comment]
I updated the jsfiddle. https://jsfiddle.net/sm706vyd/2/

**Comment by ackvf** [comment]
It can be further reduced to just this, but I find the remaining type definition helpful to showcase the issue.
```
// Incorrect

import { useWeb3React as useWeb3React_ } from '@web3-react/core'

export const useWeb3React: <T = any>(key?: string) => any

declare global {
  type SupportedChainIds = 1 | 4
}
```

```
// Correct

import { useWeb3React as useWeb3React_ } from '@web3-react/core'

export const useWeb3React: < T = any >(key?: string) => any

declare global {
  type SupportedChainIds = 1 | 4
}
```

**Comment by ackvf** [comment]
A workaround people usually use, which is accepted by the TS community is adding a comma `,` at the end of the generic, which is allowed in generics lists, but not allowed in JSX.

```
type MyFunction = <T = any>(arg: T) = ({ value: T }) // <- would cause highlighting issues
type MyOtherType = <T, U, V>(arg: T) = ({ a: T, b: U, c: V }) // <- example of correct way to use commas in type generics
```
```
type MyFunction = <T = any,>(arg: T) = ({ value: T }) // <- Accepted workaround
type MyOtherType = <T, U, V>(arg: T) = ({ a: T, b: U, c: V }) // <- no workaround needed as it already contains commas
```

**Comment by joshgoebel** [comment]
Did you glance at the PR?  See the description for the strategy we're using.

**Comment by ackvf** [comment]
Yes, you could add check for this too `,>` e.g. `<T,>` or `<T = any,>` or `<T extends string,>` or `<T extends string = 'xyz',>`, which is a generally accepted escape hatch in JSX-TS world to make parsers realise this is not JSX but TS.

**Comment by joshgoebel** [comment]
As you see from the other issue it seems like if we add the `=` case that I think now we're covering all the bases?

**Comment by ackvf** [comment]
Yeah, I guess so.
